File: docs/binary-tree/code/binary-tree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *left;
    struct Node *right;
} Node;

typedef struct {
    Node *root;
} BST;

BST create_bst() {
    BST bst;
    bst.root = NULL;
    return bst;
}

Node *create_node(int data) {
    Node *node = (Node *)malloc(sizeof(Node));
    node->data = data;
    node->left = NULL;
    node->right = NULL;
    return node;
}

Node *insert_node(Node *node, int data) {
    if (node == NULL) {
        return create_node(data);
    }

    if (data < node->data) {
        node->left = insert_node(node->left, data);
    } else if (data > node->data) {
        node->right = insert_node(node->right, data);
    }

    return node;
}

void insert(BST *bst, int data) {
    bst->root = insert_node(bst->root, data);
}

Node *search(Node *node, int data) {
    if (node == NULL) {
        return NULL;
    }

    if (data == node->data) {
        return node;
    }

    if (data < node->data) {
        return search(node->left, data);
    } else {
        return search(node->right, data);
    }
}

Node *find_successor(Node *node) {
    node = node->right;
    while (node->left != NULL) {
        node = node->left;
    }
    return node;
}

Node *delete_node(Node *node, int data) {
    if (node == NULL) {
        return NULL;
    }

    if (data < node->data) {
        node->left = delete_node(node->left, data);
    } else if (data > node->data) {
        node->right = delete_node(node->right, data);
    } else {
        if (node->left == NULL) {
            Node *temp = node->right;
            free(node);
            return temp;
        }
        if (node->right == NULL) {
            Node *temp = node->left;
            free(node);
            return temp;
        }

        Node *successor = find_successor(node);

        node->data = successor->data;

        node->right = delete_node(node->right, successor->data);
    }

    return node;
}

void delete(BST *bst, int data) {
    bst->root = delete_node(bst->root, data);
}

void preorder(Node *node) {
    if (node == NULL) {
        return;
    }

    printf("%d ", node->data);
    preorder(node->left);
    preorder(node->right);
}

void inorder(Node *node) {
    if (node == NULL) {
        return;
    }

    inorder(node->left);
    printf("%d ", node->data);
    inorder(node->right);
}

void postorder(Node *node) {
    if (node == NULL) {
        return;
    }

    postorder(node->left);
    postorder(node->right);
    printf("%d ", node->data);
}

typedef struct QueueNode {
    Node *data;
    struct QueueNode *next;
} QueueNode;

typedef struct {
    QueueNode *front;
    QueueNode *back;
    int size;
} Queue;

Queue create_bfs_queue() {
    Queue q;
    q.front = NULL;
    q.back = NULL;
    q.size = 0;
    return q;
}

void bfs_enqueue(Queue *q, Node *data) {
    QueueNode *node = (QueueNode *)malloc(sizeof(QueueNode));
    node->data = data;
    node->next = NULL;
    if (q->back == NULL) {
        q->front = node;
        q->back = node;
    } else {
        q->back->next = node;
        q->back = node;
    }
    q->size++;
}

Node *bfs_dequeue(Queue *q) {
    QueueNode *temp = q->front;
    Node *data = temp->data;
    q->front = q->front->next;
    if (q->front == NULL) {
        q->back = NULL;
    }
    free(temp);
    q->size--;
    return data;
}

int bfs_is_empty(Queue *q) {
    return q->size == 0;
}

void level_order(BST *bst) {
    if (bst->root == NULL) {
        return;
    }

    Queue q = create_bfs_queue();
    bfs_enqueue(&q, bst->root);

    while (!bfs_is_empty(&q)) {
        Node *node = bfs_dequeue(&q);
        printf("%d ", node->data);

        if (node->left != NULL) {
            bfs_enqueue(&q, node->left);
        }

        if (node->right != NULL) {
            bfs_enqueue(&q, node->right);
        }
    }
}

int max(int a, int b) {
    return a > b ? a : b;
};

int height(Node *node) {
    if (node == NULL) {
        return -1;
    }

    return 1 + max(height(node->left), height(node->right));
}


int check_balanced(Node *node) {
    if (node == NULL) {
        return 0;
    }

    int left_height = check_balanced(node->left);
    int right_height = check_balanced(node->right);

    if (left_height == -1 || right_height == -1) {
        return -1;
    }

    if (abs(left_height - right_height) > 1) {
        return -1;
    }

    return 1 + max(left_height, right_height);
}

int is_balanced(BST *bst) {
    return check_balanced(bst->root) != -1;
}

Node *lca(Node *node, int p, int q) {
    if (node == NULL) {
        return NULL;
    }

    if (p < node->data && q < node->data) {
        return lca(node->left, p, q);
    }

    if (p > node->data && q > node->data) {
        return lca(node->right, p, q);
    }

    return node;
}

int is_valid_bst(Node *node, long min, long max) {
    if (node == NULL) {
        return 1;
    }

    if (node->data <= min || node->data >= max) {
        return 0;
    }

    return is_valid_bst(node->left, min, node->data) &&
           is_valid_bst(node->right, node->data, max);
}

int is_valid(BST *bst) {
    return is_valid_bst(bst->root, LONG_MIN, LONG_MAX);
}

int main() {
    BST bst = create_bst();
    insert(&bst, 5);
    insert(&bst, 3);
    insert(&bst, 7);
    insert(&bst, 1);
    insert(&bst, 4);
    insert(&bst, 10);
    //        [5]
    //       /   \
    //     [3]   [7]
    //    /   \     \
    //  [1]   [4]   [10]

    preorder(bst.root);     // 5 3 1 4 7 10
    printf("\n");
    inorder(bst.root);      // 1 3 4 5 7 10
    printf("\n");
    postorder(bst.root);    // 1 4 3 10 7 5
    printf("\n");
    level_order(&bst);      // 5 3 7 1 4 10
    printf("\n");

    Node *result = search(bst.root, 4);
    if (result != NULL) {
        printf("Found: %d\n", result->data);    // Found: 4
    } else {
        printf("Not found\n");
    }

    Node *missing = search(bst.root, 99);
    if (missing != NULL) {
        printf("Found: %d\n", missing->data);
    } else {
        printf("Not found\n");  // Not found
    }

    printf("%d\n", height(bst.root));   // 2
    printf("%d\n", is_balanced(&bst));  // 1 (true)

    result = lca(bst.root, 1, 4);
    printf("%d\n", result->data);   // 3

    result = lca(bst.root, 1, 7);
    printf("%d\n", result->data);   // 5

    result = lca(bst.root, 3, 4);
    printf("%d\n", result->data);   // 3

    delete(&bst, 1);    // Case 1: leaf node
    //        [5]
    //       /   \
    //     [3]   [7]
    //       \     \
    //       [4]   [10]

    delete(&bst, 3);    // Case 2: one child
    //        [5]
    //       /   \
    //     [4]   [7]
    //              \
    //              [10]

    delete(&bst, 5);    // Case 3: two children
    //        [7]
    //       /   \
    //     [4]   [10]

    BST unbalanced = create_bst();
    insert(&unbalanced, 5);
    insert(&unbalanced, 3);
    insert(&unbalanced, 1);

    printf("%d\n", is_balanced(&unbalanced));   // 0 (false)

    BST valid = create_bst();
    insert(&valid, 5);
    insert(&valid, 3);
    insert(&valid, 7);
    insert(&valid, 1);
    insert(&valid, 4);

    printf("%d\n", is_valid(&valid));   // 1 (true)

    Node *root = create_node(5);
    root->left = create_node(3);
    root->left->right = create_node(9);  // 9 > 5, invalid!

    BST invalid;
    invalid.root = root;

    printf("%d\n", is_valid(&invalid));  // 0 (false)

    return 0;
}

File: docs/binary-tree/code/binary-tree.rs
struct QueueNode<T> {
    data: T,
    next: Option<Box<QueueNode<T>>>,
}

struct Queue<T> {
    front: Option<Box<QueueNode<T>>>,
    back: *mut QueueNode<T>,
    size: usize,
}

impl<T> Queue<T> {
    fn new() -> Queue<T> {
        Queue {
            front: None,
            back: std::ptr::null_mut(),
            size: 0,
        }
    }

    fn enqueue(&mut self, data: T) {
        let mut node = Box::new(QueueNode { data, next: None });

        let raw = &mut *node as *mut QueueNode<T>;

        if self.back.is_null() {
            self.front = Some(node);
            self.back = raw;
        } else {
            unsafe {
                (*self.back).next = Some(node);
            }

            self.back = raw;
        }

        self.size += 1;
    }

    fn dequeue(&mut self) -> Option<T> {
        let node = self.front.take()?;

        self.front = node.next;

        if self.front.is_none() {
            self.back = std::ptr::null_mut();
        }

        self.size -= 1;

        Some(node.data)
    }

    fn is_empty(&self) -> bool {
        self.size == 0
    }
}

struct Node {
    data: i32,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

struct BST {
    root: Option<Box<Node>>,
}

impl BST {
    fn new() -> BST {
        BST { root: None }
    }

    fn insert(&mut self, data: i32) {
        self.root = Self::insert_node(self.root.take(), data);
    }

    fn insert_node(node: Option<Box<Node>>, data: i32) -> Option<Box<Node>> {
        match node {
            None => Some(Box::new(Node {
                data,
                left: None,
                right: None,
            })),
            Some(mut current) => {
                if data < current.data {
                    current.left = Self::insert_node(current.left.take(), data);
                } else if data > current.data {
                    current.right = Self::insert_node(current.right.take(), data);
                }

                Some(current)
            }
        }
    }

    fn search(&self, data: i32) -> Option<&Node> {
        Self::search_node(self.root.as_ref(), data)
    }

    fn search_node(node: Option<&Box<Node>>, data: i32) -> Option<&Node> {
        match node {
            None => None,
            Some(current) => {
                if data == current.data {
                    Some(current)
                } else if data < current.data {
                    Self::search_node(current.left.as_ref(), data)
                } else {
                    Self::search_node(current.right.as_ref(), data)
                }
            }
        }
    }

    fn delete(&mut self, data: i32) {
        self.root = Self::delete_node(self.root.take(), data);
    }

    fn find_successor(node: &Box<Node>) -> i32 {
        let mut current = node.right.as_ref().unwrap();
        while current.left.is_some() {
            current = current.left.as_ref().unwrap();
        }
        current.data
    }

    fn delete_node(node: Option<Box<Node>>, data: i32) -> Option<Box<Node>> {
        match node {
            None => None,
            Some(mut current) => {
                if data < current.data {
                    current.left = Self::delete_node(current.left.take(), data);
                } else if data > current.data {
                    current.right = Self::delete_node(current.right.take(), data);
                } else {
                    if current.left.is_none() {
                        return current.right;
                    }
                    if current.right.is_none() {
                        return current.left;
                    }

                    let successor_data = Self::find_successor(&current);

                    current.data = successor_data;

                    current.right = Self::delete_node(current.right.take(), successor_data);
                }

                Some(current)
            }
        }
    }

    fn preorder(&self) {
        Self::preorder_node(self.root.as_ref());
    }

    fn preorder_node(node: Option<&Box<Node>>) {
        match node {
            None => return,
            Some(current) => {
                print!("{} ", current.data);
                Self::preorder_node(current.left.as_ref());
                Self::preorder_node(current.right.as_ref());
            }
        }
    }

    fn inorder(&self) {
        Self::inorder_node(self.root.as_ref());
    }

    fn inorder_node(node: Option<&Box<Node>>) {
        match node {
            None => return,
            Some(current) => {
                Self::inorder_node(current.left.as_ref());
                print!("{} ", current.data);
                Self::inorder_node(current.right.as_ref());
            }
        }
    }

    fn postorder(&self) {
        Self::postorder_node(self.root.as_ref());
    }

    fn postorder_node(node: Option<&Box<Node>>) {
        match node {
            None => return,
            Some(current) => {
                Self::postorder_node(current.left.as_ref());
                Self::postorder_node(current.right.as_ref());
                print!("{} ", current.data);
            }
        }
    }

    fn level_order(&self) {
        if self.root.is_none() {
            return;
        }

        let mut q: Queue<*const Node> = Queue::new();

        unsafe {
            q.enqueue(self.root.as_deref().unwrap() as *const Node);

            while !q.is_empty() {
                let node = &*q.dequeue().unwrap();
                print!("{} ", node.data);

                if let Some(left) = node.left.as_deref() {
                    q.enqueue(left as *const Node);
                }

                if let Some(right) = node.right.as_deref() {
                    q.enqueue(right as *const Node);
                }
            }
        }
    }

    fn height(&self) -> i32 {
        Self::height_node(self.root.as_ref())
    }

    fn height_node(node: Option<&Box<Node>>) -> i32 {
        match node {
            None => -1,
            Some(current) => {
                let left_height = Self::height_node(current.left.as_ref());
                let right_height = Self::height_node(current.right.as_ref());

                1 + left_height.max(right_height)
            }
        }
    }

    fn is_balanced(&self) -> bool {
        Self::check_balanced(self.root.as_ref()) != -1
    }

    fn check_balanced(node: Option<&Box<Node>>) -> i32 {
        match node {
            None => 0,
            Some(current) => {
                let left_height = Self::check_balanced(current.left.as_ref());
                let right_height = Self::check_balanced(current.right.as_ref());

                if left_height == -1 || right_height == -1 {
                    return -1;
                }

                if (left_height - right_height).abs() > 1 {
                    return -1;
                }

                1 + left_height.max(right_height)
            }
        }
    }

    fn lca(&self, p: i32, q: i32) -> Option<&Node> {
        Self::lca_node(self.root.as_ref(), p, q)
    }

    fn lca_node(node: Option<&Box<Node>>, p: i32, q: i32) -> Option<&Node> {
        match node {
            None => None,
            Some(current) => {
                if p < current.data && q < current.data {
                    return Self::lca_node(current.left.as_ref(), p, q);
                }

                if p > current.data && q > current.data {
                    return Self::lca_node(current.right.as_ref(), p, q);
                }

                Some(current)
            }
        }
    }

    fn is_valid(&self) -> bool {
        Self::is_valid_bst(self.root.as_ref(), i64::MIN, i64::MAX)
    }

    fn is_valid_bst(node: Option<&Box<Node>>, min: i64, max: i64) -> bool {
        match node {
            None => true,
            Some(current) => {
                let val = current.data as i64;

                if val <= min || val >= max {
                    return false;
                }

                Self::is_valid_bst(current.left.as_ref(), min, val)
                    && Self::is_valid_bst(current.right.as_ref(), val, max)
            }
        }
    }
}

fn main() {
    let mut bst = BST::new();
    bst.insert(5);
    bst.insert(3);
    bst.insert(7);
    bst.insert(1);
    bst.insert(4);
    bst.insert(10);
    //        [5]
    //       /   \
    //     [3]   [7]
    //    /   \     \
    //  [1]   [4]   [10]

    bst.preorder(); // 5 3 1 4 7 10
    println!();
    bst.inorder(); // 1 3 4 5 7 10
    println!();
    bst.postorder(); // 1 4 3 10 7 5
    println!();
    bst.level_order(); // 5 3 7 1 4 10
    println!();

    match bst.search(4) {
        Some(node) => println!("Found: {}", node.data), // Found: 4
        None => println!("Not found"),
    }

    match bst.search(99) {
        Some(node) => println!("Found: {}", node.data),
        None => println!("Not found"), // Not found
    }

    println!("{}", bst.height()); // 2
    println!("{}", bst.is_balanced()); // true

    if let Some(node) = bst.lca(1, 4) {
        println!("{}", node.data); // 3
    }

    if let Some(node) = bst.lca(1, 7) {
        println!("{}", node.data); // 5
    }

    if let Some(node) = bst.lca(3, 4) {
        println!("{}", node.data); // 3
    }

    bst.delete(1); // Case 1: leaf node
                   //        [5]
                   //       /   \
                   //     [3]   [7]
                   //       \     \
                   //       [4]   [10]

    bst.delete(3); // Case 2: one child
                   //        [5]
                   //       /   \
                   //     [4]   [7]
                   //              \
                   //              [10]

    bst.delete(5); // Case 3: two children
                   //        [7]
                   //       /   \
                   //     [4]   [10]

    let mut valid = BST::new();
    valid.insert(5);
    valid.insert(3);
    valid.insert(7);
    valid.insert(1);
    valid.insert(4);

    println!("{}", valid.is_valid()); // true

    // Manually create an invalid BST
    let mut invalid = BST::new();
    invalid.root = Some(Box::new(Node {
        data: 5,
        left: Some(Box::new(Node {
            data: 3,
            left: None,
            right: Some(Box::new(Node {
                data: 9, // 9 > 5, invalid!
                left: None,
                right: None,
            })),
        })),
        right: None,
    }));

    println!("{}", invalid.is_valid()); // false
}
